File: src/position-manager.ts
import { EventEmitter } from "node:events";
import { CONFIG } from "./config.js";
import { createLogger } from "./utils/logger.js";
import type { Database } from "./db/database.js";
import type { BankrollManager } from "./bankroll-manager.js";
import type { Signal, SimulatedTrade, TradeResult, MarketWindow } from "./types.js";

const log = createLogger("positions");

export class PositionManager extends EventEmitter {
  private openCount = 0;

  constructor(private db: Database, private bankroll: BankrollManager) {
    super();
  }

  tryOpen(signal: Signal, window: MarketWindow, isBatch = false): SimulatedTrade | null {
    // Guard: max open positions
    if (this.openCount >= CONFIG.MAX_OPEN_POSITIONS) {
      log.debug("Max open positions reached, skipping signal");
      return null;
    }

    // Guard: bankroll allows trading
    if (!this.bankroll.canTrade()) {
      return null;
    }

    // Guard: duplicate / opposing position prevention
    const existing = this.db.getOpenTradesForMarket(window.marketId);
    if (isBatch) {
      // Batch signals (e.g., spread-capture buys both sides atomically):
      // only block exact duplicates (same strategy + same direction)
      const duplicate = existing.find(
        (t) => t.strategy === signal.strategy && t.side === signal.direction,
      );
      if (duplicate) {
        log.debug(
          `Already have ${duplicate.strategy} ${duplicate.side} in ${window.marketId}, skipping duplicate`,
        );
        return null;
      }
    } else {
      // Single signals: block ANY position from the same strategy in this market
      // (prevents opposing UP+DOWN bets that guarantee a net loss)
      const sameStrategyTrade = existing.find(
        (t) => t.strategy === signal.strategy,
      );
      if (sameStrategyTrade) {
        log.debug(
          `Already have ${sameStrategyTrade.strategy} ${sameStrategyTrade.side} in ${window.marketId}, ` +
          `blocking ${signal.direction} (opposing position prevention)`,
        );
        return null;
      }
    }

    const entryPrice = signal.direction === "UP" ? signal.upAsk : signal.downAsk;
    const tokenId = signal.direction === "UP" ? window.upTokenId : window.downTokenId;

    // Bankroll-aware position sizing
    const size = this.bankroll.reserveCapital(entryPrice, signal.confidence);
    if (size <= 0) {
      log.debug(`Bankroll rejected trade: insufficient capital or no edge`);
      return null;
    }

    const trade: SimulatedTrade = {
      timestamp: signal.timestamp,
      marketId: window.marketId,
      strategy: signal.strategy,
      side: signal.direction,
      tokenId,
      entryPrice,
      size,
      status: "open",
    };

    const id = this.db.openTrade(trade);
    trade.id = id;
    this.openCount++;

    const cost = (entryPrice * size).toFixed(2);
    log.info(
      `TRADE OPENED #${id}: ${trade.strategy} ${trade.side} @ ${trade.entryPrice.toFixed(3)} ` +
      `(${size} tokens, $${cost} cost) [market: ${window.question}]`,
    );

    this.emit("tradeOpened", trade);
    return trade;
  }

  resolveWindow(window: MarketWindow, openPrice: number, closePrice: number): void {
    // BTC went UP if closePrice >= openPrice
    const outcome = closePrice >= openPrice ? "UP" : "DOWN";

    log.info(
      `Resolving window: ${window.question} | ` +
      `open=$${openPrice.toFixed(2)} close=$${closePrice.toFixed(2)} => ${outcome}`,
    );

    const trades = this.db.getOpenTradesForMarket(window.marketId);
    if (trades.length === 0) {
      log.info("No open trades to resolve for this window");
      this.db.resolveMarket(window.marketId, "resolved");
      return;
    }

    for (const trade of trades) {
      const won = trade.side === outcome;
      const settlementPrice = won ? 1.0 : 0.0;

      // P&L = (settlement - entry) * size
      const gross = (settlementPrice - trade.entryPrice) * trade.size;
      const entryCost = trade.entryPrice * trade.size;

      const result: TradeResult = {
        tradeId: trade.id!,
        exitPrice: settlementPrice,
        settlementPrice,
        pnl0pct: gross,
        pnl1pct: gross - entryCost * 0.01,
        pnl2pct: gross - entryCost * 0.02,
        pnl3pct: gross - entryCost * 0.03,
      };

      this.db.closeTrade(trade.id!, result);
      this.openCount = Math.max(0, this.openCount - 1);

      // Update bankroll
      this.bankroll.applyTradeResult(trade.id!, trade.entryPrice, trade.size, settlementPrice);

      const emoji = won ? "WIN" : "LOSS";
      log.info(
        `TRADE RESOLVED #${trade.id}: ${emoji} ${trade.strategy} ${trade.side} | ` +
        `entry=${trade.entryPrice.toFixed(3)} settlement=${settlementPrice} | ` +
        `P&L(0%)=$${result.pnl0pct.toFixed(2)} P&L(3%)=$${result.pnl3pct.toFixed(2)}`,
      );

      this.emit("tradeResolved", trade, result);
    }

    this.db.resolveMarket(window.marketId, "resolved");
  }
}

File: src/position_manager.rs
// Position manager — manages trade lifecycle (open / close / settle).
//
// Ported from the TypeScript `PositionManager` class.  Instead of extending
// `EventEmitter`, methods return values that the caller feeds to the circuit
// breaker and trend tracker.

use crate::bankroll::BankrollManager;
use crate::clock::Clock;
use crate::config::Config;
use crate::db::database::Database;
use crate::types::{
    MarketWindow, Signal, SignalDirection, SimulatedTrade, TradeResult, TradeStatus,
};

/// Manages the lifecycle of simulated (paper) trades.
///
/// Tracks how many positions are currently open and enforces the
/// `max_open_positions` limit from [`Config`].  Settlement results are
/// returned to the caller instead of being emitted via events.
pub struct PositionManager {
    open_count: u64,
}

impl Default for PositionManager {
    fn default() -> Self {
        Self::new()
    }
}

impl PositionManager {
    pub fn new() -> Self {
        Self { open_count: 0 }
    }

    /// Try to open a single position for the given signal.
    ///
    /// Returns `Some(trade)` on success, `None` if the position was blocked
    /// by any guard (max positions, duplicate strategy, insufficient capital).
    ///
    /// When `is_batch` is `true` (spread-capture), only exact duplicates
    /// (same strategy + same direction) are blocked.  When `false`, *any*
    /// existing position from the same strategy in the same market is blocked.
    #[allow(clippy::too_many_arguments)]
    pub fn try_open(
        &mut self,
        signal: &Signal,
        window: &MarketWindow,
        is_batch: bool,
        db: &Database,
        bankroll: &mut BankrollManager,
        config: &Config,
        clock: &dyn Clock,
    ) -> Option<SimulatedTrade> {
        // Guard: max open positions.
        if self.open_count >= config.max_open_positions {
            return None;
        }

        // Guard: bankroll allows trading.
        if !bankroll.can_trade(config, clock) {
            return None;
        }

        // Guard: duplicate position in the same market.
        let existing = db.get_open_trades_for_market(&window.market_id).ok()?;
        if is_batch {
            // Block exact duplicates (same strategy + same direction).
            let duplicate = existing
                .iter()
                .any(|t| t.strategy == signal.strategy && t.side == signal.direction);
            if duplicate {
                return None;
            }
        } else {
            // Block ANY position from the same strategy.
            let same_strategy = existing.iter().any(|t| t.strategy == signal.strategy);
            if same_strategy {
                return None;
            }
        }

        // Determine entry price and token ID based on direction.
        let entry_price = match signal.direction {
            SignalDirection::Up => signal.up_ask,
            SignalDirection::Down => signal.down_ask,
        };
        let token_id = match signal.direction {
            SignalDirection::Up => window.up_token_id.clone(),
            SignalDirection::Down => window.down_token_id.clone(),
        };

        // Reserve capital via the bankroll manager.
        let size = bankroll.reserve_capital(
            entry_price,
            signal.confidence,
            &signal.strategy,
            config,
            clock,
        );
        if size <= 0.0 {
            return None;
        }

        let mut trade = SimulatedTrade {
            id: None,
            timestamp: signal.timestamp,
            market_id: window.market_id.clone(),
            strategy: signal.strategy.clone(),
            side: signal.direction,
            token_id,
            entry_price,
            size,
            status: TradeStatus::Open,
        };

        let id = db.open_trade(&trade).ok()?;
        trade.id = Some(id);
        self.open_count += 1;

        Some(trade)
    }

    /// Try to open a spread (both UP and DOWN legs) for the given signals.
    ///
    /// Returns a `Vec` of exactly two trades on success, or an empty `Vec` if
    /// any guard blocks the spread.
    #[allow(clippy::too_many_arguments)]
    pub fn try_open_spread(
        &mut self,
        signals: &[Signal],
        window: &MarketWindow,
        db: &Database,
        bankroll: &mut BankrollManager,
        config: &Config,
        clock: &dyn Clock,
    ) -> Vec<SimulatedTrade> {
        // Need room for 2 more positions.
        if self.open_count + 2 > config.max_open_positions {
            return Vec::new();
        }

        if !bankroll.can_trade(config, clock) {
            return Vec::new();
        }

        // Find the UP and DOWN signals.
        let up_signal = signals.iter().find(|s| s.direction == SignalDirection::Up);
        let down_signal = signals
            .iter()
            .find(|s| s.direction == SignalDirection::Down);

        let (Some(up_signal), Some(down_signal)) = (up_signal, down_signal) else {
            return Vec::new();
        };

        // Check for existing duplicates.
        let Ok(existing) = db.get_open_trades_for_market(&window.market_id) else {
            return Vec::new();
        };

        for signal in [up_signal, down_signal] {
            let dup = existing
                .iter()
                .any(|t| t.strategy == signal.strategy && t.side == signal.direction);
            if dup {
                return Vec::new();
            }
        }

        // Reserve capital for both legs.
        let (up_tokens, down_tokens) = bankroll.reserve_spread_capital(
            up_signal.up_ask,
            down_signal.down_ask,
            up_signal.confidence,
            config,
            clock,
        );
        if up_tokens <= 0.0 || down_tokens <= 0.0 {
            return Vec::new();
        }

        let legs: [(&Signal, f64); 2] = [(up_signal, up_tokens), (down_signal, down_tokens)];
        let mut trades = Vec::with_capacity(2);

        for (signal, size) in legs {
            let entry_price = match signal.direction {
                SignalDirection::Up => signal.up_ask,
                SignalDirection::Down => signal.down_ask,
            };
            let token_id = match signal.direction {
                SignalDirection::Up => window.up_token_id.clone(),
                SignalDirection::Down => window.down_token_id.clone(),
            };

            let mut trade = SimulatedTrade {
                id: None,
                timestamp: signal.timestamp,
                market_id: window.market_id.clone(),
                strategy: signal.strategy.clone(),
                side: signal.direction,
                token_id,
                entry_price,
                size,
                status: TradeStatus::Open,
            };

            match db.open_trade(&trade) {
                Ok(id) => {
                    trade.id = Some(id);
                    self.open_count += 1;
                    trades.push(trade);
                }
                Err(_) => return Vec::new(),
            }
        }

        trades
    }

    /// Settle all open trades in a resolved market window.
    ///
    /// Returns a `Vec` of `(trade, result)` pairs.  The caller is responsible
    /// for feeding these to the circuit breaker and trend tracker.
    #[allow(clippy::too_many_arguments)]
    pub fn resolve_window(
        &mut self,
        window: &MarketWindow,
        open_price: f64,
        close_price: f64,
        db: &Database,
        bankroll: &mut BankrollManager,
        config: &Config,
        clock: &dyn Clock,
    ) -> Vec<(SimulatedTrade, TradeResult)> {
        let outcome = if close_price >= open_price {
            SignalDirection::Up
        } else {
            SignalDirection::Down
        };

        let Ok(trades) = db.get_open_trades_for_market(&window.market_id) else {
            let _ = db.resolve_market(&window.market_id, "resolved");
            return Vec::new();
        };

        if trades.is_empty() {
            let _ = db.resolve_market(&window.market_id, "resolved");
            return Vec::new();
        }

        let mut results = Vec::with_capacity(trades.len());

        for trade in trades {
            let Some(trade_id) = trade.id else {
                continue;
            };
            let won = trade.side == outcome;
            let settlement_price = if won { 1.0 } else { 0.0 };
            let gross = (settlement_price - trade.entry_price) * trade.size;
            let entry_cost = trade.entry_price * trade.size;

            let result = TradeResult {
                trade_id,
                exit_price: settlement_price,
                settlement_price,
                pnl_0pct: gross,
                pnl_1pct: gross - entry_cost * 0.01,
                pnl_2pct: gross - entry_cost * 0.02,
                pnl_3pct: gross - entry_cost * 0.03,
            };

            let _ = db.close_trade(trade_id, &result);

            if self.open_count > 0 {
                self.open_count -= 1;
            }

            bankroll.apply_trade_result(
                trade_id,
                trade.entry_price,
                trade.size,
                settlement_price,
                &trade.strategy,
                config,
                db,
                clock,
            );

            results.push((trade, result));
        }

        let _ = db.resolve_market(&window.market_id, "resolved");

        results
    }

    /// Current number of open positions.
    pub fn open_count(&self) -> u64 {
        self.open_count
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
#[path = "tests/position_manager_tests.rs"]
mod tests;
